### PR TITLE
gateway-js: Add onSchemaLoadOrUpdate() method to register listeners for schema loads/updates and give them the core supergraph SDL

### PR DESCRIPTION
**Pull request by sachindshinde**
For gateway schema reporting, we would like to know the (core) supergraph SDL when the schema is loaded or updates. To that end, this PR:
- Adds a new public method `onSchemaLoadOrUpdate()` that registers listeners for schema loads/updates.
  - Unlike `onSchemaChange()`, this method ensures listeners are called during initial schema load.
    - Specifically, `onSchemaChange()` would not notify listeners when the schema is loaded via `loadStatic()`, but `onSchemaLoadOrUpdate()` does notify.
  - Unlike `onSchemaChange()`, this method provides listeners with both the API schema and the core supergraph SDL.
  - Unlike `onSchemaChange()`, this method provides listeners data via properties on a single argument object instead of multiple arguments.
- Deprecates `onSchemaChange()` in favor of `onSchemaLoadOrUpdate()`.
- Factors out some common logic into `updateWithSchemaAndNotify()`.
- Changes the notification process to notify all listeners even when one throws (previously, any listener throwing would prevent other listeners from being notified).
- Runs prettier on `gateway-js/src/index.ts`.
- Updates the `gateway-js/CHANGELOG.md`.

### CODE REVIEW

**Comment by glasser** [comment]
Is there a high level design of "gateway schema reporting" available?

**Comment by sachindshinde** [comment]
@glasser 
There's Josh's quip doc "Schema Reporting from the Gateway", but I've been told it's out-of-date. (The rather big change is that we're not changing the API for `reportServerInfo`, and instead just passing the supergraph core SDL as the `executableSchema`.)

I've made the PR to update the plugin at apollographql/apollo-server#5187 ; that PR's description goes more into detail about the approach used.

File: gateway-js/src/index.ts
Comment (glasser) [inline gateway-js/src/index.ts]: This seems like a backwards-incompatible change (for TypeScript users) where existing Apollo Server won't be able to pass in the function it currently passes in. Though maybe I'm wrong due to how assignability works in TypeScript. Perhaps you could write a tiny test that shows that passing in a single-arg callback still compiles?

Hmm. OK, I played around and you're right that this won't break using old AS with new Gateway. But I think the other way around may be a challenge: passing a two-arg function to an old Gateway that expects one arg will not compile.  And in fact that makes sense: if you use new AS with old Gateway you will have code in your callback that expects to get two `string`s but the second one won't be there.

I suggest that we define this callback type as `(schema: GraphQLSchema, supergraphSdl?: string) => void`, and in apollo-server you check if supergraphSdl is defined and blow up with some error about needing to upgrade `@apollo/gateway` if it isn't there (or do some other behavior that is ok with no supergraphSdl).

Do you intend to eventually migrate the SchemaChangeCallback type? Leaving it around indefinitely with the wrong typing sounds confusing. Not sure exactly what the constraints here are.

**Review by glasser — CHANGES_REQUESTED** [review]

File: gateway-js/src/index.ts
Comment (glasser) [reply gateway-js/src/index.ts]: Ah you allude to this in https://github.com/apollographql/apollo-server/pull/5187

> This is dependent on apollographql/federation#738 . Specifically, when this PR lands, users will need to have access to a version of @apollo/gateway containing the referenced PR, or else their TS presumably won't compile (specifically, I believe they'll observe that ApolloGateway won't implement the interface GraphQLService, although I haven't checked this yet).

I see that that PR updates the SchemaChangeCallback type too.

I think you're right about what the problem will be. Making the second arg optional would help here.  (Or making an entirely new callback type and onSchemaChangeWithSupergraphSdl entry point! I think second arg optional is better though.)

File: gateway-js/src/index.ts
Comment (trevor-scheer) [reply gateway-js/src/index.ts:40]: Either is ok with me. I'm personally a fan of it, but the state of `prettier` in this repo (unenforced) puts me on the fence and I don't feel too strongly in any case.

File: gateway-js/src/index.ts
Comment (trevor-scheer) [reply gateway-js/src/index.ts]: Hmm, I think the bit about `load` is a fair point but our current API encourages users to specifically _not_ handle calling `load` themselves and to let AS handle that. Still totally fine to do it the "old way", but at this point I think it's undocumented.

Is there a reason you think it's bad to call the update listeners on initial load? Noisy/not semantic maybe? I can appreciate that viewpoint.

Comment (glasser) [reply gateway-js/src/index.ts]: I mean, I agree that this API is for use by AS, and AS explicitly doesn't care about getting onSchemaChange callbacks called on the initial load (its callback does nothing if the server isn't started yet). So the fact that it's for use by AS suggests that if we want to resolve the inconsistency in the current behavior (which I think is a good idea) we should do it in the direction of fewer redundant calls rather than more?

Comment (trevor-scheer) [reply gateway-js/src/index.ts]: `onSchemaChange` is not necessarily just for AS, that could reasonably be used in userland for observability (though I see now it's undocumented). We use it for promise driven tests, so we may need to update some if its behavior is changed.

I definitely agree we should make it consistent. I also agree less redundant calls is better, though I'm not convinced this is redundant (for the userland use case).

Comment (glasser) [reply gateway-js/src/index.ts]: @trevor-scheer OK, I'm back from the long weekend, let's resolve this :) Sounds like we agree it should be consistent.

My instinct is that calling the `onSchemaChange` *before `load` resolves* with the original schema seems odd to me, and not a great fit for the "change" in its name. So I lean towards not calling it for the original load. But this is an area you are more deeply involved with, so how about you choose which way to make it consistent?

File: gateway-js/src/index.ts
Comment (sachindshinde) [reply gateway-js/src/index.ts:40]: Same, I'm also a fan of `prettier`. I'd be fine with keeping the `prettier` commit in the PR, but unsquashed as @glasser  suggested. (If we do have a large rewrite pending that clashes, let me know and I can drop the commit.)

File: gateway-js/src/index.ts
Comment (sachindshinde) [reply gateway-js/src/index.ts]: I agree with @trevor-scheer here. While our own code only uses `onSchemaChange()` for schema reporting, it's part of the public API (albeit undocumented) and `load()` is expected to be called by Apollo Server, so I'd err on the side of exposing the initial schema load to `onSchemaChange` callbacks.

File: gateway-js/src/index.ts
Comment (sachindshinde) [reply gateway-js/src/index.ts]: I'm down with making the second positional argument in `SchemaChangeCallback` optional, and having the check essentially be a runtime check instead of a compile-time check. We can give better messaging with a runtime check, and we do a fair number of runtime checks already at gateway start so it's not unexpected. I've updated apollographql/apollo-server#5187 to make the argument optional.

File: gateway-js/src/index.ts
Comment (glasser) [reply gateway-js/src/index.ts]: See somewhat rambly https://github.com/apollographql/apollo-server/pull/5187#discussion_r635498292 for an idea of how to resolve this: basically, I think y'all are right that calling it always is reasonable, but let's make a more clear name for it. Coming up with a clearer name also gives us the opportunity to make this callback take an object rather than positional arguments, and to more easily detect whether the new version of the gateway is running.

(If you go through with that plan, then I think it's fine to consistently call callbacks registered either via the new name or the old name with the first schema.)

**Comment by sachindshinde** [comment]
@glasser @trevor-scheer 
Update here, following the feedback given on PRs and in Slack, this PR has been changed to:
- Not change any behavior for `onSchemaChange()` at all (i.e. still one argument for the callback, and still no notification on initial schema load for `loadStatic()`).
  - Common logic has still been refactored out though into `updateWithSchemaAndNotify()`.
- Add a new method `onSchemaLoadOrUpdate()` that:
  - Is notified on both initial schema load and later schema updates.
  - Passes a single object type to the callback instead of positional arguments.
  - Passes both the core supergraph SDL and the API schema to the callback.
- Mark `onSchemaChange()` as deprecated, in favor of `onSchemaLoadOrUpdate()`.

If this sounds fine, let me know and I'll update the PR description and `CHANGELOG.md`.

**Review by glasser — APPROVED** [review]
This looks right. Do be sure to update the PR description (and eventual squash commit) to describe what the latest version does, and the CHANGELOG as you mentioned in it.

Re the two comments below: mostly just take a glance to make sure your PR hasn't made anything worse (eg made it more likely that an error could lead to those fields getting out of sync with the actual schema), and perhaps add some TODO comments? I mean if you're inspired to introduce SchemaDerivedData I wouldn't complain but I'm sure you want to merge too.

File: gateway-js/src/index.ts
Comment (sachindshinde) [reply gateway-js/src/index.ts:481]: Yep, I'm down to move it into `updateWithSchemaAndNotify()`. We weren't doing it before in `loadStatic()`, but that's probably because it's only called at initial schema load, so it just ends up being a no-op.

I agree that "schema derived data" would be nice in gateway, although I'll leave that to future PRs.

File: gateway-js/src/index.ts
Comment (sachindshinde) [reply gateway-js/src/index.ts:403]: Hmmm, looking at the code there's two things that make me a bit wary about moving it into `updateWithSchemaAndNotify()`:
- Based on this block:
    ```typescript
    // TODO(trevor): #580 redundant parse
    // This may throw, so we'll calculate early (specifically before making any updates)
    // In the case that it throws, the gateway will:
    //   * on initial load, throw the error
    //   * on update, log the error and don't update
    const parsedSupergraphSdl = parse(result.supergraphSdl);
    ```
    It looks like we parse early to avoid accidental state update (we could double-parse, but I think we've been trying to avoid redundant parses).
- We don't update `this.parsedSupergraphSdl` in `updateByComposition()`, which initially gave me the impression we only update it when we're handed a core supergraph schema directly and not when we get a service list. But looking at `loadStatic()`, we do actually update `this.parsedSupergraphSdl` when running `this.createSchemaFromServiceList()`, so I'm not sure why we don't update it in `updateByComposition()`. Maybe @trevor-scheer might know more here?

**Comment by sachindshinde** [comment]
I've updated the PR description and the `CHANGELOG.md` for the new changes.

I won't merge this PR in until apollographql/apollo-server#5187 gets approved. (There's a somewhat lingering question above about `this.parsedSupergraphSdl`, but no behavior is changing in this PR with regards to it, so I figure we can handle it later if it's not something simple.)

**Comment by sachindshinde** [comment]
Note that there will be a slight merge-conflict with #807 (shouldn't be hard to fix, but worth noting).

**Review by glasser — APPROVED** [review]
I did a quick re-skim and this looks good!

**Review by WUTICHAI5661102493 — COMMENTED** [review]
Th

**Pull request merged**
